File: smartmoneyserver/src/main/java/fpt/aptech/server/dto/category/CategoryRequest.java
package fpt.aptech.server.dto.category;

import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.Builder;

/**
 * DTO nhận dữ liệu tạo/sửa danh mục từ Client.
 * Sử dụng Java Record kết hợp Validation.
 */
@Builder
public record CategoryRequest(
    
    @NotBlank(message = "Category name cannot be empty")
    @Size(max = 100, message = "Category name must be at most 100 characters")
    String ctgName,

    @NotNull(message = "Category type cannot be empty")
    Boolean ctgType, // true: Thu nhập, false: Chi tiêu

    @Size(max = 2048, message = "Icon URL is too long")
    String ctgIconUrl,

    // ID danh mục cha (nếu tạo danh mục con). Null nếu tạo danh mục gốc.
    Integer parentId
) {}

File: APP MOBILE/smart_money/lib/modules/category/models/category_request.dart
/// Request tạo/sửa danh mục gửi lên server.
/// Tương ứng: CategoryRequest.java (server)
class CategoryRequest {
  final String ctgName;
  final bool ctgType;
  final String? ctgIconUrl;
  final int? parentId;

  const CategoryRequest({
    required this.ctgName,
    required this.ctgType,
    this.ctgIconUrl,
    this.parentId,
  });

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'ctgName': ctgName,
      'ctgType': ctgType,
    };
    if (ctgIconUrl != null) map['ctgIconUrl'] = ctgIconUrl;
    if (parentId != null) map['parentId'] = parentId;
    return map;
  }
}

